Add a binding for OneBusAway Data

sdwilsh opened the pull request:
OneBusAway is a transit data provider used in a number of locations:
https://onebusaway.org/onebusaway-deployments/

sdwilsh commented:
One thing that I'd like to add to this still is some type of discovery.  However, I would only want to do discovery based on stops that are already added as a Thing (to add routes at that stop).  The docs all seem to assume I can discover something automatically, and I didn't see any existing examples in-tree that did something similar

sdwilsh commented:
I don't believe the jenkins failure is related to my change at all.

martinvw commented on addons/binding/org.openhab.binding.onebusaway/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,33 @@
+Bundle-Name: OneBusAway Binding
+Bundle-SymbolicName: org.openhab.binding.onebusaway;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.0.0.qualifier
```
Wrong version number here

sdwilsh replied:
Fairly certain this was generated; probably want to check that the generation script is updated.

martinvw replied:
Yes, you are right, see https://github.com/eclipse/smarthome/pull/3278 if you like a review is more than welcome!

martinvw commented on addons/binding/org.openhab.binding.onebusaway/build.properties:
```diff
@@ -0,0 +1,6 @@
+bin.includes = META-INF/,\
+               .,\
+               OSGI-INF/,\
+               ESH-INF/
```
Missing new line at the en of file.

martinvw replied:
Can you also add the about.html here

sdwilsh replied:
done

martinvw commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/config/StopConfiguration.java:
```diff
@@ -0,0 +1,60 @@
+/**
+ * Copyright (c) 2014-2016 by the respective copyright holders.
```
Wrong license year, please check all

sdwilsh replied:
These were generated this way, and I just copied the generated headers.  Probably worth making sure the generation script does the right thing (and maybe not hardcode the date so it works in the future without updating all the places since this seems to be a common problem).

Regardless, I'll fix this in all the files I've added.

martinvw commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/handler/ApiHandler.java:
```diff
@@ -0,0 +1,83 @@
+     */
+    @Override
+    public void initialize() {
+        logger.info("Initializing OneBusAway bridge...");
```
Please make it a debug, in normal operation you should not have loggin

sdwilsh replied:
done

martinvw commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/handler/RouteHandler.java:
```diff
@@ -0,0 +1,278 @@
+                    break;
+                default:
+                    logger.warn("Unhandled command for channel: {} with comamnd: {}", channelUID.getId(),
+                            command.toString());
```
The toString is automaticly called by the logging framework, you should just pass in the command the rest is handled for you

sdwilsh replied:
done

martinvw commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/handler/RouteHandler.java:
```diff
@@ -0,0 +1,278 @@
+
+    private void scheduleTriggerEvents(ChannelUID channelUID, Calendar now,
+            List<ArrivalAndDeparture> arrivalAndDepartures) {
+        scheduleTriggerEvents(channelUID, now, arrivalAndDepartures, new Function<ArrivalAndDeparture, Long>() {
```
I think our java version supports the real lambda's any reason you are using it like this

sdwilsh replied:
Honestly, I haven't used Java extensively since before that came out, and as a result I'm a bit rusty with Java, but I can update this to use them.

martinvw replied:
No problem, its a nice addition you definitely consider to read about them later, it's a not blocker here.

martinvw commented on addons/binding/org.openhab.binding.onebusaway/pom.xml:
```diff
@@ -0,0 +1,23 @@
+  <name>OneBusAway Binding</name>
+  <packaging>eclipse-plugin</packaging>
+
+  <properties>
```
I believe these are not required can you remove, I'll remove them from the others later on

sdwilsh replied:
My random sampling of other plugins has `name` and `packaging` still, but not `properties`, so I'm just removing `properties`.

martinvw commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/OneBusAwayBindingConstants.java:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Copyright (c) 2014-2016 by the respective copyright holders.
```
It seems that this file has a missing  or invalid copyright header, can you fix all of them by running 

`mvn license:format` 

Please commit only the ones for your binding.

sdwilsh replied:
Luckily it was only my binding that needed updating.  It was non-obvious that that command had to be ran from the top level though.

martinvw commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/handler/RouteHandler.java:
```diff
@@ -0,0 +1,278 @@
+     */
+    @Override
+    public void initialize() {
+        logger.info("Initializing OneBusAway route stop...");
```
Please check the other info statements

sdwilsh replied:
Updated the remaining three sites that used info.

martinvw commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/handler/RouteHandler.java:
```diff
@@ -0,0 +1,278 @@
+        }
+        StopHandler stopHandler = getStopHandler();
+        if (stopHandler != null) {
+            stopHandler.registerRouteDataListener(this);
```
The thing seems not to be marked online

sdwilsh replied:
I'm not quite sure I understand your comment.  Do you mean that it's never marked as online?  If that's what you meant, I can explain (and I'll add a comment in the code since it wasn't obvious during code review here).

The `stopHandler` here will notify this thing when it has data for the stop (it is the bridge for this thing).  The bridge is online once it loads data (see `StopHandler`), and then when we get initialized, we call `registerRouteDataListener`, which will notify this class (via `onNewRouteData`), which then marks things as online.

martinvw replied:
Yes that's what I meant. Is there a reason why you should not already mark it online, or maybe INITIALISING (or some other similar status)?

sdwilsh replied:
Setting it to `INITIALIZING` will throw an `IllegalArgumentException` with the message `Illegal status INITIALIZING. Bindings only may set UNKNOWN, ONLINE, OFFLINE or REMOVED`.

sdwilsh commented:
The newest update addresses the review comments, and fixes two other issues I found while testing it:
* The JSON format changed slightly, which caused an NPE
* The last update time was never being published.  This bug existed before, and I'm not sure how I missed it.

martinvw commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/config/ApiConfiguration.java:
```diff
@@ -0,0 +1,61 @@
+ *
+ * @author Shawn Wilsher - Initial contribution
+ */
+public class ApiConfiguration {
```
I missed on import one all packages which you do not export (and are thus not meant to be called by others) should be inside the internal package. This applies for example to the `config` and `handler` package.

See [guidelines B.10](http://docs.openhab.org/developers/development/guidelines.html#b-osgi-bundles)

sdwilsh replied:
Easy enough to fix.

martinvw commented:
Force a rebuild

martinvw commented:
Thanks for the update! I feel guilty to aks something again but like the check from pullaprove mentions, you should include the a `Signed-off-by` (to certify that you wrote it or otherwise have the right to pass it on as an open-source), see http://docs.openhab.org/developers/contributing/contributing.html#sign-your-work

Maybe the big boss has some additional remark, so for me it's okay to do it after @kaikreuzer s approval.

sdwilsh commented:
I swear I read through that document and did all the things, but I don't remember that.  I'll get that fixed up tomorrow.

I'm guessing this is your equivalent to a contributor license agreement?

martinvw commented:
> I'm guessing this is your equivalent to a contributor license agreement?

Indeed :-)

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/RouteHandler.java:
```diff
@@ -0,0 +1,292 @@
+     */
+    @Override
+    public void handleUpdate(ChannelUID channelUID, State newState) {
+        logger.warn("The OneBusAway binding is a read-only binding and can not handle channel updates.");
```
remove this line and simply ignore state updates

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/RouteHandler.java:
```diff
@@ -0,0 +1,292 @@
+                    }
+                    break;
+                default:
+                    logger.warn("Unhandled command for channel: {} with comamnd: {}", channelUID.getId(), command);
```
Should rather be something like "unknown channel UID"....

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/RouteHandler.java:
```diff
@@ -0,0 +1,292 @@
+                    logger.warn("Unhandled command for channel: {} with comamnd: {}", channelUID.getId(), command);
+            }
+        } else {
+            logger.warn("The OneBusAway binding is a read-only binding and can not handle commands.");
```
Reduce to debug - commands that are not relevant should be simply ignored.

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/RouteHandler.java:
```diff
@@ -0,0 +1,292 @@
+    @Override
+    public void initialize() {
+        logger.debug("Initializing OneBusAway route stop...");
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_PENDING, "Checking configuration...");
```
Remove this line, see my comment above.

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/RouteHandler.java:
```diff
@@ -0,0 +1,292 @@
+    private StopHandler getStopHandler() {
+        Bridge bridge = getBridge();
+        if (bridge == null || !(bridge.getHandler() instanceof StopHandler)) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR, "No bridge available");
```
Note:
1. getBridge() should always return you an instance, so it is unlikely that we get here
2. dispose() must NOT call updateStatus, but as it calls getStopHandler it potentially might do this. So you should remove this line here.

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/RouteHandler.java:
```diff
@@ -0,0 +1,292 @@
+            if (remaining.isEmpty()) {
+                return;
+            }
+            scheduledFutures.add(scheduler.schedule(new Runnable() {
```
You can use Java8 closures instead:
```
scheduledFutures.add(scheduler.schedule(() -> { publishChannel(channelUID, Calendar.getInstance(), lastUpdateTime, remaining); }
```

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/RouteHandler.java:
```diff
@@ -0,0 +1,292 @@
+            }
+
+            // Schedule this trigger
+            scheduledFutures.add(scheduler.schedule(new Runnable() {
```
use Java8

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/StopHandler.java:
```diff
@@ -0,0 +1,273 @@
+            logger.debug("Refreshing {}...", channelUID);
+            forceUpdate();
+        } else {
+            logger.warn("The OneBusAway binding is a read-only binding and can not handle commands.");
```
reduce to debug

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/StopHandler.java:
```diff
@@ -0,0 +1,273 @@
+     */
+    @Override
+    public void handleUpdate(ChannelUID channelUID, State newState) {
+        logger.warn("The OneBusAway binding is a read-only binding and can not handle channel updates.");
```
remove

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/StopHandler.java:
```diff
@@ -0,0 +1,273 @@
+    @Override
+    public void initialize() {
+        logger.debug("Initializing OneBusAway stop bridge...");
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_PENDING, "Checking configuration...");
```
remove

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/StopHandler.java:
```diff
@@ -0,0 +1,273 @@
+        }
+
+        // Do the rest of the work asynchronously because it can take a while.
+        scheduler.submit(new Runnable() {
```
use Java8

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/StopHandler.java:
```diff
@@ -0,0 +1,273 @@
+            @Override
+            public void run() {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_PENDING, "Initializing...");
+                httpClient = new HttpClient();
```
Do not instantiate and start a new HttpClient instance for every call - this is very expensive.
Keep the started client as a field within your class instance and reuse it. Also best make sure to synchronize the calls (or just ignore new calls if one is currently still running; note that `fetchAndUpdateStopData` is potentially called many times in a row, when a REFRESH is sent and processed for all channels).

sdwilsh replied:
This is creating the `HttpClient` class asynchronously when the call to `initialize` happens, but not every time.  I'm not sure where I read this at this point, but I was under the impression that expensive work should be done asynchronously and not during the `initialize` call.

I don't like synchronizing, so I'll add some logic to bail out if we are already fetching with an `AtomicBoolean`.

kaikreuzer replied:
> expensive work should be done asynchronously 

This is right, but this refers to calls that might involve remote communication that can take several seconds and potentially result in time-outs. Creating an HttpClient instance is expensive, if it is done a hundred times, but it is ok to create it in initialize() directly.

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/StopHandler.java:
```diff
@@ -0,0 +1,273 @@
+                    httpClient = null;
+                    return;
+                }
+                gson = new Gson();
```
Do not instantiate a new Gson instance for every call - this is very expensive.
Keep the instance as a field within your class instance and reuse it.

sdwilsh replied:
This is creating the `Gson` class asynchronously when the call to `initialize` happens, but not every time.  I'm not sure where I read this at this point, but I was under the impression that expensive work should be done asynchronously and not during the `initialize` call.

kaikreuzer replied:
Ok, same here: Should be fine to do it directly in initialize(), or probably even directly in the constructor (as it does not need to be closed/disposed in any way).

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/StopHandler.java:
```diff
@@ -0,0 +1,273 @@
+                    return;
+                }
+
+                scheduleJob();
```
What is this call doing within your runnable?

sdwilsh replied:
This is the deferred work of initialization.  We don't want to schedule our updates until we know we've been successfully been initialized.

kaikreuzer replied:
Once you have moved the creation of HttpClient and Gson out of the runnable, there is no other code left that then scheduleJob - so you can completely remove the runnable and simply call scheduleJob (which then in turn should run immediately the first time).

And yes, we WANT to schedule regular updates, since we want to keep trying to establish a connection. If the first request fails, the second one might actually succeed. If you do not schedule the polling right away, you will never be able to recover.

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/StopHandler.java:
```diff
@@ -0,0 +1,273 @@
+            pollingJob.cancel(true);
+            pollingJob = null;
+        }
+        scheduler.execute(pollingRunnable);
```
why do you schedule it once when you schedule it continuously in the next line?

sdwilsh replied:
This is `forceUpdate`; I want to update the data right now.  `scheduleJob` only schedules the polling in the future.  I should clearly rename that method to make this more clear.

I don't run it immediately in `scheduleJob` because when we are initializing ourselves, we want to make sure we can load data before we mark ourselves online.  We already have our data at that point, so it would be wasteful to then make another request if `scheduleJob` would run immediately and then scheduled the updates.  In `forceUpdate`, we also want to reset the update schedule since we just got new data (hence the canceling of the job).

kaikreuzer replied:
I would reduce this method to simply `scheduler.execute(pollingRunnable);`.
This method should do a fetch "out of the row", I would not change the existing polling job in here - this isn't done in any other binding either and it makes the overall logic much simpler and better understandable.

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/StopHandler.java:
```diff
@@ -0,0 +1,273 @@
+    }
+
+    private void scheduleJob() {
+        assert (pollingJob == null);
```
what is this assert good for?

sdwilsh replied:
My own sanity while developing.  I can remove it.

kaikreuzer commented on addons/binding/org.openhab.binding.onebusaway/src/main/java/org/openhab/binding/onebusaway/internal/handler/StopHandler.java:
```diff
@@ -0,0 +1,273 @@
+
+    private void scheduleJob() {
+        assert (pollingJob == null);
+        pollingJob = scheduler.scheduleWithFixedDelay(pollingRunnable, config.getInterval(), config.getInterval(),
```
You should schedule it to run immediately the first time.

sdwilsh replied:
See above; this would be a waste of local resources and API requests (they have a limit too) since in order to know we are online, we have to make a data request anyway.  We also don't want to schedule our re-occurring jobs until we know we are online, as that would waste local resources running a job which is doomed to fail.

kaikreuzer commented:
Oops, one thing I just noticed: the README.md seems to be missing in this PR, correct?

sdwilsh commented on addons/binding/org.openhab.binding.onebusaway/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,45 @@
+            <label>API Key</label>
+            <description>The OneBusAway API key to use.</description>
+        </parameter>
+        <parameter name="apiServer" type="text" required="true">
```
I could attempt to construct a list, but the various agencies put this information in different places, unfortunately.

kaikreuzer replied:
Maybe at least list the most popular ones - you can set limitToOptions=false, so the user can also enter any other server that he would like to use.

sdwilsh commented on addons/binding/org.openhab.binding.onebusaway/ESH-INF/thing/api.xml:
```diff
@@ -0,0 +1,13 @@
+        xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+    <bridge-type id="api">
+        <label>OneBusAway API</label>
```
It's literally an API key, and each server admin (for different transit regions) has different policies on how to hand these out, as far as I can tell.  In the pugent sound region (Seattle, WA), you just email them and they give you an API key.

kaikreuzer replied:
No, this is not the "API key" here, this is the "bridge" (which needs an API key as a configuration). So the "Thing" that requires the key is the web service - that's why I was asking for a more speaking name.

sdwilsh commented on addons/binding/org.openhab.binding.onebusaway/ESH-INF/thing/route.xml:
```diff
@@ -0,0 +1,27 @@
+        <supported-bridge-type-refs>
+            <bridge-type-ref id="stop" />
+        </supported-bridge-type-refs>
+        <label>OneBusAway Route</label>
```
Bus is a misnomer; this can and is used for things like light rail in a subway, for instance.

kaikreuzer replied:
Then just call it "Route". As it is a route of a bus/subway/railway, but not that of a OneBusAway.

sdwilsh commented on addons/binding/org.openhab.binding.onebusaway/ESH-INF/thing/stop.xml:
```diff
@@ -0,0 +1,16 @@
+        <supported-bridge-type-refs>
+            <bridge-type-ref id="api" />
+        </supported-bridge-type-refs>
+        <label>OneBusAway Stop</label>
```
Bus is a misnomer; this can and is used for things like light rail in a subway, for instance.

kaikreuzer replied:
See above - just call it "Stop".

sdwilsh commented on addons/binding/org.openhab.binding.onebusaway/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,54 @@
+    <channel-type id="arrival">
+        <item-type>DateTime</item-type>
+        <label>Arrival Time</label>
+        <description>The arrival time of the event</description>
```
I went with event based in part with how the astro binding named things, and because you can add an offset on the channel, so it's not quite the arrival time, but rather time of the event as you have configured it.

The word "bus" is a misnomer; this can and is used for things like light rail in a subway, for instance.

I'm open to a better name, but I haven't come up with one.

kaikreuzer replied:
It is the arrival time of the bus/subway/railway (and you can configure an offset, yes). So if "bus" is the wrong word, simply write "The arrival time" here.

sdwilsh commented on addons/binding/org.openhab.binding.onebusaway/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,54 @@
+    <channel-type id="departure">
+        <item-type>DateTime</item-type>
+        <label>Departure Time</label>
+        <description>The departure time of the event</description>
```
I went with event based in part with how the astro binding named things, and because you can add an offset on the channel, so it's not quite the departure time, but rather time of the event as you have configured it.

The word "bus" is a misnomer; this can and is used for things like light rail in a subway, for instance.

I'm open to a better name, but I haven't come up with one.

kaikreuzer replied:
See above.

sdwilsh commented on addons/binding/org.openhab.binding.onebusaway/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,54 @@
+        <label>Last Update Time</label>
+        <description>The last time information was updated</description>
+        <state readOnly="true" pattern="%1$tF %1$tR" />
+        <config-description-ref uri="channel-type:onebusaway:config" />
```
I wasn't going to at first, but then I thought that it could be useful to write a rule that runs if, say, the information was last updated more then five minutes ago.  I'm fine dropping this, but adding the flexibility (and making all these channels work the same way) was easy to support.

kaikreuzer replied:
>  to write a rule that runs if, say, the information was last updated

This is a feature that should be part of the rule engine (this is what https://github.com/eclipse/smarthome/pull/3241 tries to do), not of (every single) channel - so I would strongly suggest to remove it here.

sdwilsh commented:
> Oops, one thing I just noticed: the README.md seems to be missing in this PR, correct?

I thought I had read somewhere that this was auto-generated by the bindings.  I'm guessing that is incorrect.  Is there a guide on how this should be structured?

kaikreuzer commented:
Here's how the README.md should be done: https://www.eclipse.org/smarthome/documentation/development/bindings/docs.html
Also simply check the README.md of other bindings for reference.

martinvw commented:
@sdwilsh you are aware that we still need a README?

sdwilsh commented:
I'm aware; life just got really busy so I haven't had time to circle back on this yet.  I should get to it soon, however.

kaikreuzer commented:
@sdwilsh Just the readme missing - don't let us down ;-)

sdwilsh commented:
I'm in the process of starting a major home renovation so my free time has been basically zero.  I'm pretty sure I'll get back to this before July though :)

kaikreuzer commented:
I'd like to include it in the upcoming openHAB 2.1 release (due in around 10 days) - so if you could simply choose a night to sleep 30 minutes less than usual and write the readme, we could have it in the release and you would have an item less on your todo list for the next weeks. Doesn't that sound appealing? 😆

sdwilsh commented:
I plan to look at this today.  I just re-based locally, and I want to test it to make sure it still works correctly.

sdwilsh commented:
I found and fixed a few bugs when I tested this with a clean slate.  I should have addressed all your review comments at this time as well.
